studio/data/profile/mfa-challenge-and-verify-mutation.ts

```diff
@@ -0,0 +1,48 @@
+import { AuthMFAVerifyResponse, MFAChallengeAndVerifyParams } from '@supabase/supabase-js'
+import { useMutation, UseMutationOptions, useQueryClient } from '@tanstack/react-query'
+import { auth } from 'lib/gotrue'
+import { profileKeys } from './keys'
+
+interface MFAChallengeAndVerifyVariables extends MFAChallengeAndVerifyParams {
+  refreshFactors?: boolean
+}
+
+export const mfaChallengeAndVerify = async (params: MFAChallengeAndVerifyParams) => {
+  const { error, data } = await auth.mfa.challengeAndVerify(params)
+  if (error) throw error
+  return data
+}
+
+type CustomMFAVerifyResponse = NonNullable<AuthMFAVerifyResponse['data']>
+type CustomMFAVerifyError = NonNullable<AuthMFAVerifyResponse['error']>
+
+export const useMfaChallengeAndVerifyMutation = ({
+  onSuccess,
+  ...options
+}: Omit<
+  UseMutationOptions<CustomMFAVerifyResponse, CustomMFAVerifyError, MFAChallengeAndVerifyVariables>,
+  'mutationFn'
+> = {}) => {
+  const queryClient = useQueryClient()
+
+  return useMutation(
+    (vars) => {
+      const { refreshFactors, ...params } = vars
+      return mfaChallengeAndVerify(params)
+    },
+    {
+      async onSuccess(data, variables, context) {
+        // when a MFA is added, the aaLevel is bumped up
+        const refreshFactors = variables.refreshFactors ?? true
+
+        await Promise.all([
+          ...(refreshFactors ? [queryClient.invalidateQueries(profileKeys.mfaFactors())] : []),
```

Comment:
cc @ivasilov just fyi i added this change here - mainly so when signing in we don't need to refresh this (just saves us one call)

Reply:
@joshenlim unfortunately, this addition is made irrelevant by this line https://github.com/supabase/supabase/blob/15818116b9c5a1ceb7d924a02139f458b89ec3c1/studio/components/interfaces/SignIn/SignInMfaForm.tsx#L78. All queries gets reset (and they need to be refetched) which causes the input field to clear. It bugged me too so I had to go down the rabbit hole.

Reply:
ahh i see hmmmm... i think we might still be able to keep this in

am thinking what could've been happening is like after challengeAndVerify succeeds, it invalidates mfafactors + aa levels, and then re-invalidates those 2 again because of the resetQueries

so am thinking its possibly helping to reduce duplicate requests?